File: protocol/src/main/resources/csharp/ProtocolManagerTemplate.cs
using System;
using System.Collections.Generic;

namespace zfoocs
{
    public class ProtocolManager
    {
        public static readonly short MAX_PROTOCOL_NUM = short.MaxValue;


        private static readonly IProtocolRegistration[] protocols = new IProtocolRegistration[MAX_PROTOCOL_NUM];
        private static readonly Dictionary<Type, short> protocolIdMap = new Dictionary<Type, short>();


        public static void InitProtocol()
        {
            {}
        }

        public static IProtocolRegistration GetProtocol(short protocolId)
        {
            var protocol = protocols[protocolId];
            if (protocol == null)
            {
                throw new Exception("[protocolId:" + protocolId + "] not exist");
            }

            return protocol;
        }

        public static void Write(ByteBuffer buffer, object packet)
        {
            var protocolId = protocolIdMap[packet.GetType()];
            // 写入协议号
            buffer.WriteShort(protocolId);

            // 写入包体
            GetProtocol(protocolId).Write(buffer, packet);
        }

        public static object Read(ByteBuffer buffer)
        {
            var protocolId = buffer.ReadShort();
            return GetProtocol(protocolId).Read(buffer);
        }
    }
}

File: protocol/src/main/resources/swift/ProtocolManagerTemplate.swift
import Foundation

class ProtocolManager {
    static var protocols = Dictionary<Int, IProtocolRegistration>()

    static func initProtocol() {
        // initProtocol
        ${protocol_manager_registrations}
    }
    
    static func getProtocol(_ protocolId: Int) -> IProtocolRegistration {
        return protocols[protocolId]!
    }
    
    static func write(_ buffer: ByteBuffer, _ packet: Any) {
        let p = packet as! IProtocol
        let protocolId = p.protocolId()
        let pro = getProtocol(protocolId)
        buffer.writeShort(Int16(protocolId))
        pro.write(buffer, p)
    }
    
    static func read(_ buffer: ByteBuffer) -> Any {
        let protocolId = buffer.readShort()
        let pro = getProtocol(Int(protocolId))
        return pro.read(buffer)
    }
}
